### PR TITLE
build: migrate to hercules ci, haskell.nix, and upgrade to ghc 8.8.4

### PR DESCRIPTION
**Pull request by brendanhay**
Continues the work started in #3374 (GHC 8.8.4, no nixcrpkgs, etc.) with additional fixes atop the `release/next-vere` branch. In addition to various Linux/Darwin static compilation issues, a number of KH specific warnings needed to be either fixed or suppressed.

Additionally:

* The urbit/vere version is now located outside of the `pkg/urbit/configure` script in `pkg/urbit/version`. This allows CI or other tooling to simply `cat version`.
* CI attributes are dimensioned by `<system>` - so we get top-level `linux` and `darwin` sets of builds currently. See https://hercules-ci.com/github/urbit/urbit/jobs/1367
* An attempt has been made to ensure `libsigsegv` defines `HAVE_STACKVMA 0` for the currently supported platforms.
* The `<system>.release` attribute contains an IFD that pushes the statically linked binaries (in a tarball) to `bootstrap.urbit.org/ci`. For example: `gs://bootstrap.urbit.org/ci/urbit-v0.10.8-x86_64-linux-<sha256>.tar.gz`.
* All pills are {re,}built whenever `/sys` changes, by virtue of Nix inputs.
* Nearly all of the `builder.sh` use has been nuked. Primarily because it subverts all of the existing Nix build machinery including various fixup phases and makes overriding/extending derivations which use the pattern problematic.
* The cargo culting of `rec { ... }` has been eliminated.
* Some of the `sh/` scripts that were no longer applicable have been removed.
* Code formatters for nix/shell have been added under `sh/fmt` and `sh/fmt-*`. Currently only the `fmt-nix` script will modify the files in place.

There are many build or CI optimisations that could be added on top of this work, such as replacing some of the IFDs  (import-from-derivations) for pushing tarballs with Hercules' nearly released effect system.

I recommend taking a look at `nix/default.nix`, `nix/overlays`, `default.nix`, and `ci.nix`. I've tried to include instructions and motivation where appropriate.

Fixes #3380 #3374 #3353 #2537

### CODE REVIEW

**Review by joemfb — CHANGES_REQUESTED** [review]
I'm not able to build this on MacOS -- I get the following error:

```
$ nix-shell --show-trace
error: while evaluating the attribute 'buildInputs' of the derivation 'urbit' at /Users/joemfb/Projects/urbit/pkg/urbit/shell.nix:6:3:
while evaluating the attribute 'src' of the derivation 'ivory-header' at /Users/joemfb/Projects/urbit/nix/pkgs/pill/ivory.nix:25:5:
while evaluating anonymous function at /Users/joemfb/Projects/urbit/nix/lib/fetch-github-lfs.nix:3:1, called from /Users/joemfb/Projects/urbit/nix/pkgs/pill/ivory.nix:6:9:
attribute 'isPath' missing, at /Users/joemfb/Projects/urbit/nix/lib/fetch-github-lfs.nix:9:8
```

I haven't been able to find anything about that error, any pointers?

---

Leaving aside my build problem, this looks pretty clean to me. I've read through the files you've mentioned and more, but I'm anything but a nix expert. It doesn't look like this will change much about the ways most of us interact with the monorepo during development. Here's what I note (some of which we've already discussed):

- New dependencies will be added in `nix/sources.json`, or vendor'd in `pkgs` as before. They'll still need derivations in `nix/pkgs`, but no more separate cross-compilation derivations.
- Derivations can be limited to a particular platform (as in `nix/lib/test-fake-ship.nix`)
- We'll need to get used to hercules-ci. I assume it builds both direct pushes to branches/tags of and pull requests to urbit/urbit (although I haven't been able to verify that yet).
- Release artifacts will need to be manually/externally promoted to `bootstrap.urbit.org` with the appropriate name.
- Tarballs no longer include an outer directory, so install/getting-started docs should be updated to account for that.

Is there anything else that's a notable change from a development/ops standpoint?

---

And finally, a typo:

**Comment by joemfb** [comment]
One more thing:

> All pills are {re,}built whenever /sys changes, by virtue of Nix inputs.

I don't see this predicate specifically -- do you mean that pills are built whenever `sys/` changes by virtue of the fact that they depend on `pkg/arvo`, and are built whenever it changes?

These rebuilt pills are uploaded to GCP *and* used for any subsequent local ship boots, most notably, the arvo tests. Is that correct? (They're obviously not committed back into the repo via LFS.)

File: sh/update-ivory-pill [hunk 1]
```diff
@@ -1,6 +1,7 @@
-#!/usr/bin/env bash
+##!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (brendanhay) [reply sh/update-ivory-pill:1]: Bizarre, thanks.

**Comment by brendanhay** [comment]
The nix error is saying that `builtins.isPath` doesn't exist for you - what `nix --version` are you running? It's just a sanity check so I'll remove it in the interests of backwards compatibility.

> New dependencies will be added in nix/sources.json, or vendor'd in pkgs as before. They'll still need derivations in nix/pkgs, but no more separate cross-compilation derivations.

You can also use `niv add <name> <url>` and subsequently `niv update <name>` to avoiding dealing with the sha256 and so on, which is how I originally generated the `sources.json` (niv is part of //shell.nix), but it's also fine to do it by hand.

>  We'll need to get used to hercules-ci. I assume it builds both direct pushes to branches/tags of and pull requests to urbit/urbit 

That's correct - it'll evaluate _everything_ with the advantage being that any branch pushed with a minor change will only build that change since the rest of the expressions inputs will be identical to any prior built branch.

It's probably worth highlighting that Hercules first evaluates the expressions starting at the root of `ci.nix` and walks the tree of attributes which due to laziness requires `recurseIntoAttrs` or our `dimension` helper to 'tag' those attributes we want exposed on CI. This evaluation occurs on a Linux machine `agent01` and once completed the attributes are filtered to the systems they support and then built by the machine which supports those features: `agent01` for Linux x86_64 and `agen02` for Darwin x86_64, currently.

When passing `enableStatic: true` on Linux you will get cross compilation to a Musl64 target and on Darwin it'll best effort statically link everything up until `libSystem`.

> Release artifacts will need to be manually/externally promoted to bootstrap.urbit.org with the appropriate name. 

I'll add a `sh/` script to use `gsutil` to do the copy in place.

> Tarballs no longer include an outer directory, so install/getting-started docs should be updated to account for that.

This is a mistake on my part - I'll re-add the outer directory. I'd also remove the `ci/` prefix from the bucket too upon merge unless it's a useful change.

> Is there anything else that's a notable change from a development/ops standpoint?

Some of the `.gitignore` and `.gitattributes` have been changed due to how Nix (and Hercules) fetches tarballs from GitHub to populate the build - this was resulting in things that we didn't want in an export but require for a build not being unpacked/present on CI.

I also have some concerns about KH work being done outside of a Nix shell - stack in no way guarantees hermicity or reproducibility for people developing on OSX, such as native dependencies being manually installed or whatever. I've provided a `pkg/hs/shell.nix` in case people want to use it, as otherwise there's no real certainty that if you code builds locally it'll build on the Darwin CI agent.

> I don't see this predicate specifically -- do you mean that pills are built whenever sys/ changes by virtue of the fact that they depend on pkg/arvo, and are built whenever it changes?

That's correct. There are `<pill>.lfs`, `<pill>.build` and `<pill>.header` attributes - everything pretty much depends on `.lfs` except for the fakeship, tests, and push-to-bucket attributes on CI. The `sh/` scripts relating to update/build pills calls the `.build` attribute for the appropriate pill and can be overridden by setting `withRopsten = true` to obtain ropsten pills.

You can choose what to depend on locally - it should be `.lfs` by default but I'll just go and verify that right now ..

If arvo has changed then CI will push a new `<pill>-<sha256>.pill` to `bootstrap.urbit.org/ci` - this might be unnecessary? It was more of an exploration of what not having pills in git might look like. 🙂

**Comment by ohAitch** [comment]
Note #3820, when functional, will either step on the toes of some of this or aid it considerably: briefly, it introduces a "crystal" slim-pill mechanism, which depends only on hoon.hoon - and furthermore, depends on the least recent hoon.hoon capable of compiling the current one to fixedpoint, so is probably best to leave near-hardcoded. (It technically also depends on some fairly trivial nock currently embodied in a rewritten `gen/glass`, which as far as CI is concerned is even more static.)

The problem of "obviously not committed back into the repo via LFS" can then be sidestepped entirely by building non-minimal pills up from there, and distributing them via cachix to speed up tests. They do after all contain the entire arvo file tree and are thus as joe said dependent on it.

Optionally keeping the GCP upload, though like, `+solid` is only really needed on release(which granted should happen through special CI handling of release tags), and a `+brass`-equivalent full-system `+glass` pill can be built from a stripped one in under 30 seconds. (Which like `+brass` you pay for by having to compile at boot time.)

**Comment by brendanhay** [comment]
@ohAitch Sounds great - I'm on board with anything that solves the issues outlined in Joe's [urbit-dev] "pill management needs to be reconsidered" post. I'm happy to make any requisite changes to the Nix expressions if/when required.

(My only concern is crystal is 7 characters and not 5.)

**Comment by brendanhay** [comment]
Your comment about cachix reminds me to add:

Now that _everything_ is built via Nix it is _all_ cached. Likely this will require paying for more cache storage but IMO it's worth it - people who aren't frequently touching arvo could in fact use the `.build` attributes of the desired pill to pull it straight from the cache. Same goes for any other build artefact including any of the Haskell code/libraries.

If someone pushes a PR that builds something expensive - when that is merged to master or another branch we no longer have to potentially pay for that build twice.

**Comment by brendanhay** [comment]
Maybe also of interest is how I configure Nix to distribute builds:

```
# /etc/nixos/configuration.nix or whatever your nix.conf is:
nix = {
  distributedBuilds = true;

  buildMachines = [
    { 
      hostName = "build";
      system = "x86_64-linux";
      maxJobs = 24;
    }
    { 
      hostName = "agent02";
      system = "x86_64-darwin";
      maxJobs = 4;
    }
  ];
}
```

```
# .ssh/config
Host build
...

Host agent02
...
```

Which transparently builds the code on the appropriate build host rather than my puny wee laptop. Note that I'm actually using Darwin CI as a personal build agent. :trollface: 

Since CI and a build farm are as Joe Bryan would say, 'isomorphic' - we could in theory just have everyone using remote builders/substitutions to maximise sharing/caching.

**Comment by brendanhay** [comment]
The release tarballs are now archived/pushed as:

```
url: gs://bootstrap.urbit.org/ci/urbit-v0.10.9-rc1-x86_64-linux-218e21[...].tgz

content:
urbit-v0.10.9-rc1-x86_64-linux/urbit
urbit-v0.10.9-rc1-x86_64-linux/urbit-king
urbit-v0.10.9-rc1-x86_64-linux/urbit-worker
```

This the difference being the `{linux,darwin64}` suffix since I've simply interpolated the Nix system' name which will be:

* `x86_64-darwin`
* `x86_64-linux`
* `aarch64-linux`
* ...

@joemfb Should I create a mapping to translate those Nix systems into the previous `linux64`, `darwin64` style suffix?

I also considered truncating the SHA256 to 128 bits to shorten the object key slightly. 🤷🏽‍♂️

File: pkg/hs/stack.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-resolver: lts-14.21
+resolver: lts-16.15
```
Comment (eglaysher) [inline pkg/hs/stack.yaml:1]: yessssssss

**Review by eglaysher — COMMENTED** [review]
Are there additional things that need to be done to hook up the caching, and should they be part of the documentation? I tried building a release tarball on linux and it started building GHC. I'm trying to enter a `nix-shell` on OSX and it's building Cabal.

File: nix/pkgs/hs/default.nix [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        "text"
+        "time"
+        "transformers"
+        "unix"
+        "xhtml"
+      ];
+
+      packages = {
+        # Disable the urbit-king test-suite for now - since it relies
+        # on relative paths to lfs pills.
```
[the remark this answers is not in the document]
Comment (brendanhay) [reply nix/pkgs/hs/default.nix:67]: Nice catch - I typically prefix these kind of things with `FIXME` so I catch them before merging, I was going to revisit this.

**Comment by brendanhay** [comment]
> Are there additional things that need to be done to hook up the caching, and should they be part of the documentation? I tried building a release tarball on linux and it started building GHC. I'm trying to enter a `nix-shell` on OSX and it's building Cabal.

I don't have access to the current cachix so I set up a new one:

```
cachix use urbit-linux
```

We can change this once someone throws the credentials at me.

**Comment by eglaysher** [comment]
Are there additional things that I need to do to get static builds on linux?

```
$ nix-build -A tarball --arg enableStatic true
[...]
/nix/store/y5z72fhq4fdxy898jvaka0mjfyh1yr6a-urbit-v0.10.9-rc1-x86_64-linux.tgz
$ tar xvf /nix/store/y5z72fhq4fdxy898jvaka0mjfyh1yr6a-urbit-v0.10.9-rc1-x86_64-linux.tgz
urbit-v0.10.9-rc1-x86_64-linux/urbit
urbit-v0.10.9-rc1-x86_64-linux/urbit-king
urbit-v0.10.9-rc1-x86_64-linux/urbit-worker
$ ldd urbit
	not a dynamic executable
$ ldd urbit-worker
	not a dynamic executable
$ ldd urbit-king
	linux-vdso.so.1 (0x00007fff02ef8000)
	libm.so.6 => /nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib/libm.so.6 (0x00007f408a681000)
	libpthread.so.0 => /nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib/libpthread.so.0 (0x00007f408a660000)
	librt.so.1 => /nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib/librt.so.1 (0x00007f408a656000)
	libutil.so.1 => /nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib/libutil.so.1 (0x00007f408a651000)
	libdl.so.2 => /nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib/libdl.so.2 (0x00007f408a64a000)
	libc.so.6 => /nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib/libc.so.6 (0x00007f408a489000)
	libnuma.so.1 => /nix/store/88190q6lqz30kg7zih3jyb4sln4vdf0w-numactl-2.0.14/lib/libnuma.so.1 (0x00007f408a47a000)
	/nix/store/bdf8iipzya03h2amgfncqpclf6bmy3a1-glibc-2.32/lib/ld-linux-x86-64.so.2 => /nix/store/xg6ilb9g9zhi2zg1dpi4zcp288rhnvns-glibc-2.30/lib64/ld-linux-x86-64.so.2 (0x00007f408a7c6000)
```

(I still haven't done any testing on OSX.)

**Comment by brendanhay** [comment]
@eglaysher Looks like `enableStatic` isn't being correctly threaded on Darwin then - I'll investigate.

**Comment by brendanhay** [comment]
The `nix-build ci.nix -A linux.tarball` appears to work as expected but `default.nix` doesn't. The former would require a faux service-account.json to exist to run it though.

I'll see if I can figure out what's up.

**Comment by brendanhay** [comment]
@eglaysher Does `nix-build -A tarball --argstr crossSystem x86_64-unknown-linux-musl` work?

**Comment by eglaysher** [comment]
After starting that command, I appear to now be building a full musl toolchain.

**Comment by joemfb** [comment]
I was on nix-2.2.1; upgrading to nix-2.3.8 fixed the `isPath` issue. I can now build the project and enter the nix-shell.

There's a pretty frequent linker warning:

```
ld: warning: passed two min versions (10.12.0, 10.12) for platform macOS. Using 10.12.
```

And an old problem with the tests has returned:

```
tail: error writing 'standard output': Resource temporarily unavailable
```

**Comment by eglaysher** [comment]
I can confirm that specifying crossSystem above actually did make a static urbit-king binary.

**Comment by brendanhay** [comment]
@joemfb Looks related to https://github.com/NixOS/nixpkgs/issues/101330 - I'll take a look tomorrow and see if there's a fix for the ld warning.

Any context for the tail error? I haven't encountered this before, I assume this is happening when you run the tests in a shell or via top-level attribute?

**Comment by joemfb** [comment]
IIRC, it's `EAGAIN` wrt `stdout`. I think we fixed this in travis-ci like so: https://github.com/travis-ci/travis-ci/issues/4704#issuecomment-348435959

**Comment by joemfb** [comment]
I also recall reading https://serverfault.com/questions/369757/why-is-this-tee-losing-stdout, and maybe removing a `tee` invocation from the test runner.

**Comment by ohAitch** [comment]
> @\ohAitch Sounds great - I'm on board with anything that solves the issues outlined in Joe's [urbit-dev] "pill management needs to be reconsidered" post. 

Hah looks like I had a reply to that half-written up I never finished, oops. Most of it now is implemented in `ohAitch/clearpill`, but it does contain the questions:

> A solid pill can be thought of as a partially evaluated brass pill, the result of all the bootstrap computation that doesn't directly depend on unique identity and entropy (ie, the symmetry-breaking of the full bootstrap sequence).
- How difficult would it be for a solid pill to _formally be_ a partially evaluated brass pill? I'm imagining a world here where we throw the compiler formula with `~>` during early boot so that a solid-creation process can keep it around for jet reattachment; similarly an ivory pill "snapshot-me" marker can be embedded in the bootstrap sequence, at which point the vere booting it can snapshot the pristine arvo noun to a file if it so desires. 

>  \- build-system integration such that the ivory pill is built before vere is compiled (needed for codegen)
- Could we link Vere with `stub_ivory.h` containing the proverbial "pointer to a segfault"? How much of ivory.pill is used purely to get as far as running enough hoon to generate a real one.

--- 

"Alternate staging suggestion" that I guess bounced around my head for a year before solidifying into `+glass`

- `urbit-worker` boots off `slim.pill`, which has the barest skeleton of an arvo interface necessary to accept an `-A` format list of files
- `slim.pill` looks for specific files in `sys/`, produces a formal brass boot sequence, compiling the current(and non-dehydrated) `hoon.hoon`
- `brass.pill` is saved to disk, and booted into from "scratch" (same road same nouns)
- when prompted, the initial vaneless arvo is saved to `ivory.pill`

**Comment by joemfb** [comment]
For posterity, this PR fixes #3369.

**Comment by brendanhay** [comment]
The King Haskell tests have been re-enabled and slightly reworked to allow local developers to use the LFS pill under `./bin` and have the path [configurable under CI](https://hercules-ci.com/accounts/github/urbit/derivations/%2Fnix%2Fstore%2Fg9ji2hskbjq8pyglx0y196f50kr7b13c-urbit-king-test-urbit-king-tests-0.10.8-check.drv/log?via-job=931b38cd-30c6-4b57-828d-cefe8cefeef8). Additionally the previous problematic use of `cd`ing up a few directories in the tests to set the current directory has been replaced with Cabal's `data-files` machinery.

Outstanding issues:

* Ensure intuitive use of `crossSystem` + `enableStatic` flags for `default.nix`. 
* Figure out the `EAGAIN` issue with `stdout`.

I've re-enabled the equivalent of `make test` on the OSX build agent while I attempt to debug the latter issue.

**Comment by brendanhay** [comment]
I've completed the last two issues I was aware of - let me know if there's anything still outstanding.

**Review by eglaysher — APPROVED** [review]
That's it from me. Congratulations on the big switchover.

**Review by joemfb — APPROVED** [review]
Awesome work, LGTM

**Pull request merged**